test/model/typing_status_test.dart

```diff
@@ -0,0 +1,207 @@
+      checkNotifiedOnce();
+      model.handleTypingEvent(eg.typingEvent(topicNarrow, TypingOp.start, eg.fourthUser.userId));
+      checkNotifiedOnce();
+      checkTypists({
+          dmNarrow: [eg.otherUser],
+          groupNarrow: [eg.thirdUser],
+          topicNarrow: [eg.fourthUser]});
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.stop, eg.thirdUser.userId));
+      checkTypists({dmNarrow: [eg.otherUser], topicNarrow: [eg.fourthUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.stop, eg.otherUser.userId));
+      checkTypists({topicNarrow: [eg.fourthUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(topicNarrow, TypingOp.stop, eg.fourthUser.userId));
+      checkTypists({});
+      checkNotifiedOnce();
+    });
+  });
+
+  group('cancelling old timer', () {
+    test('when typing stopped early', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({dmNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+      check(async.pendingTimers).length.equals(1);
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.stop, eg.otherUser.userId));
+      checkTypists({});
+      checkNotifiedOnce();
+      check(async.pendingTimers).isEmpty();
+    }));
+
+    test('when typing repeatedly started', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+      check(async.pendingTimers).length.equals(1);
+
+      // The new timer should be active and the old timer should be cancelled.
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      check(async.pendingTimers).length.equals(1);
+      checkNotNotified();
+    }));
+  });
+
+  group('typing start expiry period', () {
+    test('repeated typing start event resets the timer', () => awaitFakeAsync((async) async {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.start, eg.otherUser.userId));
+      checkNotifiedOnce();
+
+      async.elapse(const Duration(seconds: 10));
+      checkTypists({dmNarrow: [eg.otherUser]});
+      // We expect the timer to restart from the event.
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.start, eg.otherUser.userId));
+      checkNotNotified();
+
+      async.elapse(const Duration(seconds: 10));
+      checkTypists({dmNarrow: [eg.otherUser]});
+      checkNotNotified();
+
+      async.elapse(const Duration(seconds: 5));
+      checkTypists({});
+      checkNotifiedOnce();
+    }));
+
+
+    test('typist is removed when the expiry period ends', () => awaitFakeAsync((async) async {
```

Comment:
nit: put this before the other expiry test case — the narrative makes most sense when this one is read first, because the other is basically this plus an extra wrinkle